### PR TITLE
Ridge regressor

### PR DESCRIPTION
**Pull request by PicoCentauri**
Adding a Ridge class for linear models.

# Idea

The core of the `Ridge` class is that all inputs are `TensorMap` objects. This holds for the training data `X` the target values `y` but also for the regularization strength `alpha` and the `sample_weights`. We this the Ridge class looks fairly simple and clean in my view. 

For the construction of `y` we already have a helper function as introduced in #3. For `alpha` I will write a helper function in this PR.

# TODO
- [x] Finish valdiation function `_validate_data` and `_validate_params` 
- [x] Add tests
- [x] Add example
- [x] Add function for creating the regularization `TensorMap`.

# Acknowledgement

Thanks to @rosecers for the inspiration in #2. The problem with #2 that we can not use `sklearn` in equisolve. We have to use different alphas for values, gradients etc...

Thanks also to @Luthaf  for the linear_model in equistore-examples!

### CODE REVIEW

File: src/equisolve/numpy/models/__init__.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# -*- Mode: python; tab-width: 4; indent-tabs-mode:nil; coding:utf-8 -*-
+#
+# Copyright (c) 2022 Authors and contributors
+# (see the AUTHORS.rst file for the full list of names)
+#
+# Released under the GNU Public Licence, v3 or any higher version
+# SPDX-License-Identifier: GPL-3.0-or-later
```
Comment (Luthaf) [inline src/equisolve/numpy/models/__init__.py:7]: What? The repo says MIT, and the rest of the ecosystem is BSD =)

Comment (PicoCentauri) [reply src/equisolve/numpy/models/__init__.py:7]: Let me change this to BSD if this is our license

Comment (Luthaf) [reply src/equisolve/numpy/models/__init__.py:7]: I don't think we need to include this boilerplate in every files, but I would be in favor of changing the licence of this repo to BSD. This should be done in a separate PR though!

File: src/equisolve/utils/metrics.py [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+# (see the AUTHORS.rst file for the full list of names)
+#
+# Released under the BSD 3-Clause "New" or "Revised" License
+# SPDX-License-Identifier: BSD-3-Clause
+import numpy as np
+from equistore import TensorMap
+
+
+def rmse(y_true: TensorMap, y_pred: TensorMap) -> float:
+    """TODO: Needs to be a tensormap implementation."""
```
Comment (PicoCentauri) [inline src/equisolve/utils/metrics.py:13]: @agoscinski ;-)

File: src/equisolve/numpy/utils.py [hunk 3]
```diff
@@ -32,3 +35,16 @@ def block_to_array(block: TensorBlock, parameter_keys: List[str]):
             data = block.gradient(parameter).data
             M.append(data.reshape(np.prod(data.shape[:-1]), data.shape[-1]))
     return np.vstack(M)
+
+def tensor_map_to_dict(tensor_map: TensorMap):
```
Comment (Luthaf) [inline src/equisolve/numpy/utils.py:39]: There should be a comment in there to remove this function once https://github.com/lab-cosmo/equistore/issues/94 is implemented

Comment (agoscinski) [reply src/equisolve/numpy/utils.py:104]: added

File: src/equisolve/numpy/models/linear_model.py [hunk 4]
```diff
@@ -0,0 +1,257 @@
...
+                    sw_arr.shape == y_arr.shape
+                ), f"shapes = {sw_arr.shape} and {y_arr.shape}"
+            else:
+                sw_arr = np.ones((len(y_arr),))
+
+            w = self._numpy_lstsq_solver(X_arr, y_arr, sw_arr, alpha_arr)
+            self.coef_.append(w.reshape(-1, 1))
+
+        # Convert alpha to a dictionary to be used in external models.
+        self.alpha = tensor_map_to_dict(self.alpha)
```
Comment (PicoCentauri) [inline src/equisolve/numpy/models/linear_model.py:191]: I think this is not a good idea, because if we call the fit method twice it will raise an error the second time. A better way would be to save this in a private variable like `_alpha`. What do you say @agoscinski ?

File: src/equisolve/numpy/models/linear_model.py [hunk 5]
```diff
@@ -0,0 +1,257 @@
...
+        """Fit Ridge regression model to each block in X.
+
+        :param X: training data
+        :param y: target values
+        :param sample_weight: sample weights
+        """
+        self._validate_data(X, y)
+        self._validate_params(X, sample_weight)
+
+        for i_block, X_block in enumerate(X.blocks()):
```
Comment (DavideTisi) [inline src/equisolve/numpy/models/linear_model.py:163]:
```suggestion
        for key, X_block in X:
```

File: src/equisolve/numpy/models/linear_model.py [hunk 6]
```diff
@@ -0,0 +1,257 @@
...
+            y_block = y.block(i_block)
```
Comment (DavideTisi) [inline src/equisolve/numpy/models/linear_model.py:164]:
```suggestion
            y_block = y.block(key)
```

File: src/equisolve/numpy/models/linear_model.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+            alpha_block = self.alpha.block(i_block)
```
Comment (DavideTisi) [inline src/equisolve/numpy/models/linear_model.py:165]:
```suggestion
            alpha_block = self.alpha.block(key)
```

File: src/equisolve/numpy/models/linear_model.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+
+            # y_arr has shape lentgth of n_targets
+            y_arr = block_to_array(y_block, self.parameter_keys)[:, 0]
+
+            # alpha_arr has length of n_properties
+            alpha_arr = alpha_block.values[0]
+
+            # Sample weights
+            if sample_weight is not None:
+                sw_block = sample_weight.block(i_block)
```
Comment (DavideTisi) [inline src/equisolve/numpy/models/linear_model.py:178]:
```suggestion
                sw_block = sample_weight.block(key)
```

File: src/equisolve/numpy/models/linear_model.py [hunk 9]
```diff
@@ -0,0 +1,257 @@
...
+
+        if len(X.blocks()) != len(self.coef_):
+            raise ValueError(
+                f"Number of blocks in X ({len(X.blocks())}) does not agree "
+                f"with the number of fitted weights ({len(self.coef_)})."
+            )
+        self.alpha = dict_to_tensor_map(self.alpha)
+
+        blocks = []
+        for i_block, X_block in enumerate(X.blocks()):
```
Comment (DavideTisi) [inline src/equisolve/numpy/models/linear_model.py:213]:
```suggestion
        for key, X_block in X:
```

nd then correct `i_block` in the loop

File: src/equisolve/utils/metrics.py (hunk 2 above)
Comment (DavideTisi) [reply src/equisolve/utils/metrics.py:13]: For now cannot just do:
- check if the metadata are the same 
- ``` return np.sqrt(np.mean((y_true.values - y_pred.values) ** 2))```
- do the same for the grad
It works only if the metadata are in the same order but it is simple

Am I not getting something?

**Review by DavideTisi — COMMENTED** [review]
I still have to check the tests

File: src/equisolve/numpy/models/linear_model.py [hunk 10]
```diff
@@ -0,0 +1,257 @@
...
+
+    def _numpy_lstsq_solver(self, X, y, sample_weights, alphas):
+        # Convert problem with regularization term into an equivalent
+        # problem without the regularization term
+        num_targets, num_properties = X.shape
+        regularization_all = np.hstack((sample_weights, alphas))
+        regularization_eff = np.diag(np.sqrt(regularization_all))
+        X_eff = regularization_eff @ np.vstack((X, np.eye(num_properties)))
+        y_eff = regularization_eff @ np.hstack((y, np.zeros(num_properties)))
+        least_squares_output = np.linalg.lstsq(X_eff, y_eff, rcond=1e-13)
```
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply src/equisolve/numpy/models/linear_model.py:147]: Yes, wen can do this.

File: src/equisolve/utils/metrics.py (hunk 2 above)
Comment (PicoCentauri) [reply src/equisolve/utils/metrics.py:13]: Yes, more or less. I will write the function now...

File: src/equisolve/utils/convert.py [hunk 11]
```diff
@@ -69,21 +69,19 @@ def properties_to_tensormap(
...
         values=np.asarray(values).reshape(-1, 1),
-        samples=samples,
+        samples=Labels(["structure"], np.arange(n_samples).reshape(-1, 1)),
         components=[],
         properties=Labels(["property"], np.array([(0,)])),
     )
 
     if positions_gradients is not None:
-        if n_properties != len(positions_gradients):
+        if n_samples != len(positions_gradients):
```
[the remark this answers is not in the document]
Comment (DavideTisi) [reply src/equisolve/utils/convert.py:82]: maybe i do not understand what `M_i` is, or how the matrix positions_gradients is structured

File: tests/utils/test_metrics.py [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+        X_block.add_gradient(
+            parameter="positions",
+            data=X_gradient_data,
+            samples=position_gradient_samples,
+            components=[Labels(["direction"], np.arange(3).reshape(-1, 1))],
+        )
+
+        X = TensorMap(Labels.single(), [X_block])
+
+        assert_equal(rmse(X, X, parameter_key="positions"), [0.0])
```
Comment (DavideTisi) [inline tests/utils/test_metrics.py:61]: add a test to check what happens when you use a `parameter_key` not equal to `sample`, `positions` or `cell` (or in general when it is equal to something not allowed), since it is not explicitly checked in the `rmse` function but left to the call to the derivative parameter.

File: tests/numpy/models/test_linear_model.py [hunk 13]
```diff
@@ -0,0 +1,517 @@
...
+        )
+
+        clf = Ridge(parameter_keys=parameter_keys, alpha=alpha)
+        clf.fit(X=X, y=y)
+
+        assert_allclose(clf.coef_[0], w_exact.reshape(-1, 1), atol=1e-15, rtol=1e-8)
+
+        # Test prediction
+        X_pred = clf.predict(X)
+
```
Comment (DavideTisi) [inline tests/numpy/models/test_linear_model.py:382]: to check if two TensorMap are equal with also the same metadata you can use the operations.all_close(X_pred,y) function.

**Review by DavideTisi — COMMENTED** [review]
i think the predict is better to do with the dot among `TensorMap`. I did a commit to do that, do you like the new version?
if so maybe we want to have coeff_ only in its TensorMap form and then all the tests must be changed accordingly.

File: src/equisolve/utils/convert.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (PicoCentauri) [reply src/equisolve/utils/convert.py:82]: Maybe this is a bit unclear here indeed. The function takes a list of values (List[floats]) and gradients as list of numpy arrays (`List[np.ndarray]`). Therefore, the check we perform here is correct. We check that the number of provided list entries for `positions_gradients` and `values` are the same. This has nothing to do with how samples are stored in equistore.

I will adjust this in the code make it clear.
